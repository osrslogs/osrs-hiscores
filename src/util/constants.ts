import { BountyHunterType, Mode, SkillName, BossName, ClueType, ActivityName } from '../types';

export const BOUNTY_HUNTER: BountyHunterType[] = ['hunter', 'rogue'];

export const CLUES: ClueType[] = ['all', 'beginner', 'easy', 'medium', 'hard', 'elite', 'master'];

export const BOSSES: BossName[] = [
  'abyssalSire',
  'alchemicalHydra',
  'barrowsChests',
  'bryophyta',
  'callisto',
  'cerberus',
  'chambersOfXeric',
  'chambersOfXericChallengeMode',
  'chaosElemental',
  'chaosFanatic',
  'commanderZilyana',
  'corporealBeast',
  'crazyArchaeologist',
  'dagannothPrime',
  'dagannothRex',
  'dagannothSupreme',
  'derangedArchaeologist',
  'generalGraardor',
  'giantMole',
  'grotesqueGuardians',
  'hespori',
  'kalphiteQueen',
  'kingBlackDragon',
  'kraken',
  'kreeArra',
  'krilTsutsaroth',
  'mimic',
  'nightmare',
  'obor',
  'sarachnis',
  'scorpia',
  'skotizo',
  'theGauntlet',
  'theCorruptedGauntlet',
  'theatreOfBlood',
  'thermonuclearSmokeDevil',
  'tzKalZuk',
  'tzTokJad',
  'venenatis',
  'vetion',
  'vorkath',
  'wintertodt',
  'zalcano',
  'zulrah',
];

export const SKILLS: SkillName[] = [
  'overall',
  'attack',
  'defence',
  'strength',
  'hitpoints',
  'ranged',
  'prayer',
  'magic',
  'cooking',
  'woodcutting',
  'fletching',
  'fishing',
  'firemaking',
  'crafting',
  'smithing',
  'mining',
  'herblore',
  'agility',
  'thieving',
  'slayer',
  'farming',
  'runecraft',
  'hunter',
  'construction',
];

export const ACTIVITIES: ActivityName[] = ['leaguePoints', ...BOUNTY_HUNTER, ...CLUES, 'lastManStanding', ...BOSSES];

export const MODES: Mode[] = ['normal', 'ironman', 'ultimate', 'hardcore', 'deadman', 'seasonal', 'tournament'];
