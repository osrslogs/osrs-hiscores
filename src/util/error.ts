/* eslint-disable max-classes-per-file */

/**
 * A Error with data details. `data` will containing the data
 * that made the error occur.
 */
class DataError extends Error {
  data: string;

  constructor(msg: string, data: string) {
    super(msg);
    this.data = data;
  }
}

/**
 * Error used when a csv has unexpected structure, e.g: One entry too many / too little.
 * The raw csv can be found in the `data` property.
 */
export class InvalidCsvError extends DataError {
  constructor(csv: string) {
    super('Invalid csv structure', csv);
  }
}

/**
 * Error used when a html page has unexpected structure, e.g: Could not parse element.
 * The raw html can be found in the `data` property.
 */
export class InvalidHtmlError extends DataError {}

/**
 * Validation error for player name
 */
export class InvalidPlayerError extends Error {
  constructor(player: string) {
    super(`Invalid player name: ${player}`);
  }
}

/**
 * A generic HTTP error, default status 500.
 * Used to map HTTP status codes to errors.
 */
export class HttpError extends Error {
  status: number;

  constructor(msg: string, status = 500) {
    super(msg);
    this.status = status;
  }
}

export class NotFoundError extends HttpError {
  constructor(msg: string) {
    super(msg, 404);
  }
}

export class ServiceUnavailableError extends HttpError {
  constructor(msg: string) {
    super(msg, 503);
  }
}
