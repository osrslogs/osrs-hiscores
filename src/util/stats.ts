import { RankedStats } from '../types';

/**
 * Finds the name of the first ranked stat in rankedStats
 *
 * @param {RankedStats} rankedStats
 *
 * @returns {string | undefined} The name of the ranked stat, else undefined
 */
// eslint-disable-next-line import/prefer-default-export
export const getRankedStat = (rankedStats: RankedStats): string | undefined => {
  // eslint-disable-next-line no-restricted-syntax
  for (const [key, value] of Object.entries(rankedStats)) {
    if (value.rank > 0) {
      return key;
    }
  }
  return undefined;
};
