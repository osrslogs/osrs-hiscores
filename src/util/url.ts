import { Mode, SkillName, ActivityName } from '../types';
import { SKILLS, ACTIVITIES } from './constants';

const urlBase = 'https://secure.runescape.com';
const urlStats = 'index_lite.ws';
const urlScores = 'overall.ws';
const urlMode: { [key: string]: string } = {
  normal: 'm=hiscore_oldschool',
  ironman: 'm=hiscore_oldschool_ironman',
  ultimate: 'm=hiscore_oldschool_ultimate',
  hardcore: 'm=hiscore_oldschool_hardcore_ironman',
  deadman: 'm=hiscore_oldschool_deadman',
  seasonal: 'm=hiscore_oldschool_seasonal',
  tournament: 'm=hiscore_oldschool_tournament',
};

/**
 * Returns the correct url mode
 *
 * @param {Mode} [mode] The game mode to select. Defaults to `normal`
 *
 * @returns {string} The urlMode property
 */
const getUrlModeWithDefault = (mode: Mode = 'normal'): string => urlMode[mode];

/**
 * Builds the url string for personal stats
 *
 * @param {string} player The player name to lookup
 * @param {Mode} [mode] The game mode to select. Defaults to `normal`
 *
 * @returns {string}
 */
export const buildStatsUrl = (player: string, mode?: Mode): string => {
  const urlStatsPlayer = urlStats.concat(`?player=${player}`);
  return [urlBase, getUrlModeWithDefault(mode), urlStatsPlayer].join('/');
};

/**
 * Builds the url string for skill page
 *
 * @param {SkillName} skill The skill name to lookup
 * @param {Mode} [mode] The mode to select. Defaults to `normal`
 *
 * @returns {string}
 */
export const buildSkillPageUrl = (skill: SkillName, mode?: Mode): string => {
  const tableIdx = SKILLS.indexOf(skill);
  const urlScoresTable = urlScores.concat(`?table=${tableIdx}`);
  return [urlBase, getUrlModeWithDefault(mode), urlScoresTable].join('/');
};

/**
 * Builds the url string for activity page
 *
 * @param {ActivityName} activity The activity name to lookup
 * @param {Mode} [mode] The mode to select. Defaults to `normal`
 *
 * @returns {string}
 */
export const buildActivityPageUrl = (activity: ActivityName, mode?: Mode): string => {
  const tableIdx = ACTIVITIES.indexOf(activity);
  const urlScoresTable = urlScores.concat(`?category_type=1&table=${tableIdx}`);
  return [urlBase, getUrlModeWithDefault(mode), urlScoresTable].join('/');
};

/**
 * Builds the url string for formatted player name (highlighted in table)
 *
 * @param {string} player The player to lookup
 * @param {SkillName} [skill] The skill table to use. Defaults to `overall`
 * @param {Mode} [mode] The mode to select. Defaults to `normal`
 *
 * @returns {string}
 */
export const buildSkillPageWithDisplayNameUrl = (player: string, skill?: SkillName, mode?: Mode): string => {
  const defaultOrSkill = skill || 'overall';
  const pageUrl = buildSkillPageUrl(defaultOrSkill, mode);
  return pageUrl.concat(`&user=${player}`);
};
