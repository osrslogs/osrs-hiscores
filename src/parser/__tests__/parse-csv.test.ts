import parseCsv from '../parse-csv';
import { InvalidCsvError } from '../../util/error';
import csvConfig from '../../util/test-util/csv-mock.json';

describe('', () => {
  it('returns stats when given valid data', () => {
    const result = parseCsv(csvConfig.valid);
    expect(result).toMatchSnapshot();
  });

  it('throws CsvError when given data with NaN values', () => {
    expect(() => parseCsv(csvConfig.invalidNaN)).toThrow(InvalidCsvError);
  });

  it('throws CsvError when given data with missing skill', () => {
    expect(() => parseCsv(csvConfig.invalidMissingSkill)).toThrow(InvalidCsvError);
  });

  it('throws CsvError when given data with missing activity', () => {
    expect(() => parseCsv(csvConfig.invalidMissingActivity)).toThrow(InvalidCsvError);
  });

  it('throws CsvError when given data with missing newline EO', () => {
    expect(() => parseCsv(csvConfig.invalidMissingNewlineEOF)).toThrow(InvalidCsvError);
  });

  it('throws CsvError when given data with empty element in skill', () => {
    expect(() => parseCsv(csvConfig.invalidEmptyEntry)).toThrow(InvalidCsvError);
  });

  it('errors include raw csv data', () => {
    const expected = csvConfig.invalidEmptyEntry;
    try {
      parseCsv(expected);
    } catch (err) {
      expect(err.data).toBe(expected);
    }
  });
});
