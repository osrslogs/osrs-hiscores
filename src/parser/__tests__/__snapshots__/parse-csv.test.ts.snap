// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`returns stats when given valid data 1`] = `
Object {
  "bosses": Object {
    "abyssalSire": Object {
      "rank": 3957,
      "score": 910,
    },
    "alchemicalHydra": Object {
      "rank": 29495,
      "score": 136,
    },
    "barrowsChests": Object {
      "rank": 62120,
      "score": 178,
    },
    "bryophyta": Object {
      "rank": 42086,
      "score": 1,
    },
    "callisto": Object {
      "rank": 42086,
      "score": 1,
    },
    "cerberus": Object {
      "rank": 42086,
      "score": 1,
    },
    "chambersOfXeric": Object {
      "rank": 11149,
      "score": 456,
    },
    "chambersOfXericChallengeMode": Object {
      "rank": -1,
      "score": -1,
    },
    "chaosElemental": Object {
      "rank": 4218,
      "score": 157,
    },
    "chaosFanatic": Object {
      "rank": 33567,
      "score": 30,
    },
    "commanderZilyana": Object {
      "rank": 12540,
      "score": 283,
    },
    "corporealBeast": Object {
      "rank": 47194,
      "score": 10,
    },
    "crazyArchaeologist": Object {
      "rank": 72282,
      "score": 6,
    },
    "dagannothPrime": Object {
      "rank": 5876,
      "score": 1084,
    },
    "dagannothRex": Object {
      "rank": 7254,
      "score": 1110,
    },
    "dagannothSupreme": Object {
      "rank": 6762,
      "score": 1006,
    },
    "derangedArchaeologist": Object {
      "rank": -1,
      "score": -1,
    },
    "generalGraardor": Object {
      "rank": 29833,
      "score": 240,
    },
    "giantMole": Object {
      "rank": 4105,
      "score": 1100,
    },
    "grotesqueGuardians": Object {
      "rank": 41134,
      "score": 30,
    },
    "hespori": Object {
      "rank": 21976,
      "score": 45,
    },
    "kalphiteQueen": Object {
      "rank": 3275,
      "score": 660,
    },
    "kingBlackDragon": Object {
      "rank": 43920,
      "score": 187,
    },
    "kraken": Object {
      "rank": 29327,
      "score": 1663,
    },
    "kreeArra": Object {
      "rank": 10155,
      "score": 271,
    },
    "krilTsutsaroth": Object {
      "rank": 21394,
      "score": 89,
    },
    "mimic": Object {
      "rank": 21394,
      "score": 89,
    },
    "nightmare": Object {
      "rank": -1,
      "score": -1,
    },
    "obor": Object {
      "rank": -1,
      "score": -1,
    },
    "sarachnis": Object {
      "rank": -1,
      "score": -1,
    },
    "scorpia": Object {
      "rank": 9421,
      "score": 103,
    },
    "skotizo": Object {
      "rank": 75113,
      "score": 11,
    },
    "theCorruptedGauntlet": Object {
      "rank": -1,
      "score": -1,
    },
    "theGauntlet": Object {
      "rank": -1,
      "score": -1,
    },
    "theatreOfBlood": Object {
      "rank": 13681,
      "score": 1,
    },
    "thermonuclearSmokeDevil": Object {
      "rank": 47124,
      "score": 37,
    },
    "tzKalZuk": Object {
      "rank": 1356,
      "score": 2,
    },
    "tzTokJad": Object {
      "rank": 9792,
      "score": 17,
    },
    "venenatis": Object {
      "rank": 79082,
      "score": 1,
    },
    "vetion": Object {
      "rank": 49716,
      "score": 1,
    },
    "vorkath": Object {
      "rank": 9046,
      "score": 1160,
    },
    "wintertodt": Object {
      "rank": 29515,
      "score": 525,
    },
    "zalcano": Object {
      "rank": -1,
      "score": -1,
    },
    "zulrah": Object {
      "rank": 14328,
      "score": 1750,
    },
  },
  "bountyHunter": Object {
    "hunter": Object {
      "rank": -1,
      "score": -1,
    },
    "rogue": Object {
      "rank": -1,
      "score": -1,
    },
  },
  "clues": Object {
    "all": Object {
      "rank": 3794,
      "score": 1347,
    },
    "beginner": Object {
      "rank": -1,
      "score": -1,
    },
    "easy": Object {
      "rank": 2358,
      "score": 500,
    },
    "elite": Object {
      "rank": 8956,
      "score": 84,
    },
    "hard": Object {
      "rank": 16509,
      "score": 300,
    },
    "master": Object {
      "rank": 5902,
      "score": 63,
    },
    "medium": Object {
      "rank": 10413,
      "score": 400,
    },
  },
  "lastManStanding": Object {
    "rank": -1,
    "score": -1,
  },
  "leaguePoints": Object {
    "rank": -1,
    "score": -1,
  },
  "skills": Object {
    "agility": Object {
      "experience": 15585112,
      "level": 99,
      "rank": 1808,
    },
    "attack": Object {
      "experience": 177313466,
      "level": 99,
      "rank": 150,
    },
    "construction": Object {
      "experience": 16511799,
      "level": 99,
      "rank": 313,
    },
    "cooking": Object {
      "experience": 200000000,
      "level": 99,
      "rank": 255,
    },
    "crafting": Object {
      "experience": 30019718,
      "level": 99,
      "rank": 430,
    },
    "defence": Object {
      "experience": 21122418,
      "level": 99,
      "rank": 13518,
    },
    "farming": Object {
      "experience": 68900329,
      "level": 99,
      "rank": 656,
    },
    "firemaking": Object {
      "experience": 30088535,
      "level": 99,
      "rank": 1921,
    },
    "fishing": Object {
      "experience": 22861094,
      "level": 99,
      "rank": 2702,
    },
    "fletching": Object {
      "experience": 30073339,
      "level": 99,
      "rank": 1127,
    },
    "herblore": Object {
      "experience": 30063274,
      "level": 99,
      "rank": 265,
    },
    "hitpoints": Object {
      "experience": 86966857,
      "level": 99,
      "rank": 1247,
    },
    "hunter": Object {
      "experience": 13709772,
      "level": 99,
      "rank": 9585,
    },
    "magic": Object {
      "experience": 21975963,
      "level": 99,
      "rank": 11903,
    },
    "mining": Object {
      "experience": 28154835,
      "level": 99,
      "rank": 786,
    },
    "overall": Object {
      "experience": 1004888190,
      "level": 2277,
      "rank": 482,
    },
    "prayer": Object {
      "experience": 23607177,
      "level": 99,
      "rank": 223,
    },
    "ranged": Object {
      "experience": 39330316,
      "level": 99,
      "rank": 13324,
    },
    "runecraft": Object {
      "experience": 20053883,
      "level": 99,
      "rank": 953,
    },
    "slayer": Object {
      "experience": 20293993,
      "level": 99,
      "rank": 6756,
    },
    "smithing": Object {
      "experience": 30010168,
      "level": 99,
      "rank": 829,
    },
    "strength": Object {
      "experience": 23860522,
      "level": 99,
      "rank": 23274,
    },
    "thieving": Object {
      "experience": 22091405,
      "level": 99,
      "rank": 2469,
    },
    "woodcutting": Object {
      "experience": 32294215,
      "level": 99,
      "rank": 2015,
    },
  },
}
`;
